src/adc.rs

```diff
@@ -0,0 +1,201 @@
+    /// Code that attempts to call this method when the peripheral is already
+    /// enabled will not compile.
+    ///
+    /// Consumes this instance of `ADC` and returns another instance that has
+    /// its `State` type parameter set to [`Enabled`].
+    ///
+    /// # Limitations
+    ///
+    /// This assumes that the clock is running at 500 kHz for the calibration
+    ///
+    /// # Examples
+    ///
+    /// Please refer to the [module documentation] for a full example.
+    ///
+    /// [`Disabled`]: ../init_state/struct.Disabled.html
+    /// [`Enabled`]: ../init_state/struct.Enabled.html
+    /// [module documentation]: index.html
+    pub fn enable(self, clock: &AdcClock, syscon: &mut syscon::Handle) -> ADC {
+        syscon.enable_clock(&self.adc);
+        syscon.power_up(&self.adc);
+
+        // Start calibration
+        // The clock needs to be at 500 kHz for this task
+        self.adc.ctrl.write(|w| {
+            unsafe { w.clkdiv().bits(clock.caldiv) };
+            w.calmode().set_bit()
+        });
+
+        // Wait until the calibration is done
+        while self.adc.ctrl.read().calmode().bit_is_set() {}
+
+        self.adc
+            .ctrl
+            .write(|w| unsafe { w.clkdiv().bits(clock.div) });
+
+        ADC {
+            adc: self.adc,
+            _state: init_state::Enabled(()),
+        }
+    }
+}
+
+impl ADC<init_state::Enabled> {
+    /// Disable the ADC
+    ///
+    /// This method is only available, if `ADC` is in the [`Enabled`] state.
+    /// Code that attempts to call this method when the peripheral is already
+    /// disabled will not compile.
+    ///
+    /// Consumes this instance of `ADC` and returns another instance that has
+    /// its `State` type parameter set to [`Disabled`].
+    ///
+    /// [`Enabled`]: ../init_state/struct.Enabled.html
+    /// [`Disabled`]: ../init_state/struct.Disabled.html
+    pub fn disable(
+        self,
+        syscon: &mut syscon::Handle,
+    ) -> ADC<init_state::Disabled> {
+        syscon.disable_clock(&self.adc);
+
+        ADC {
+            adc: self.adc,
+            _state: init_state::Disabled,
+        }
+    }
+}
+
+impl<State> ADC<State> {
+    /// Return the raw peripheral
+    ///
+    /// This method serves as an escape hatch from the HAL API. It returns the
+    /// raw peripheral, allowing you to do whatever you want with it, without
+    /// limitations imposed by the API.
+    ///
+    /// If you are using this method because a feature you need is missing from
+    /// the HAL API, please [open an issue] or, if an issue for your feature
+    /// request already exists, comment on the existing issue, so we can
+    /// prioritize it accordingly.
+    ///
+    /// [open an issue]: https://github.com/lpc-rs/lpc8xx-hal/issues
+    pub fn free(self) -> pac::ADC0 {
+        self.adc
+    }
+}
+
+impl<PIN> OneShot<ADC, u16, PIN> for ADC
+where
+    PIN: Channel<ADC, ID = u8>,
+{
+    type Error = ();
+
+    // TODO Nonblocking
```

Comment:
Suggestion: Remove this TODO, get this merged, leave further work to follow-up PRs.